File: bayer_pipeline.py
#!/usr/bin/env python3
"""OV7670 Raw Bayer 处理管线 (bayer_pipeline.py)

把本项目实验得到的完整取证结论固化为可复用的纯函数管线:
  g_map                  : 6-bit 位序解码 (b1=MSB, 序 (1,7,6,5,4,3)) -> 0..252
  fix_dead_pixels        : 散点死点修复 (同色 4-邻域中值, 差>阈 替换)
  fix_defect_region      : 楔形区域缺陷修复 (带外同相位参考插值, 逐像素判定)
  render                 : 去马赛克 + WB + 伽马 -> PNG

取证结论 (docs/ 记录, 本文件固化):
- 位序: 原始字节 v=129*b7+64*b6+32*b5+16*b4+8*b3+2*b1, b0==b7, b2≈噪声;
  g(v)=128*b1+64*b7+32*b6+16*b5+8*b4+4*b3, b1 为真 MSB (空间相关 0.863 最高、
  空间连续性 0.670 最高、邻列 MAD 42.7 < 原始字节 70.4)。
- 240x320 模式: every-2nd-PCLK 采集, 每行 320 字节, RGGB 拜耳格式。
- 白平衡: 灰度世界 R/B 增益。
- 伽马: 0.85 提亮。

CLI 在 main() 内, __main__ 守卫 -> 纯函数层可被 unittest import。
"""

import argparse

import numpy as np

from bayer_demosaic import demosaic_bayer


def fix_dead_pixels(cfa, threshold=60.0, skip=None):
    """散点死点修复: 与同色 4-邻域 (上下左右, 间隔 2) 中值差 > threshold 则替换。

    skip: (h,w) bool mask, True 的像素不处理 (如区域缺陷修复区)。
    返回 (修复后的副本, 替换像素数)。
    """
    cfa = cfa.astype(np.float64).copy()
    h, w = cfa.shape
    if skip is None:
        skip = np.zeros((h, w), dtype=bool)
    count = 0
    for y in range(2, h - 2):
        for x in range(2, w - 2):
            if skip[y, x]:
                continue
            nbrs = []
            for dy, dx in ((-2, 0), (2, 0), (0, -2), (0, 2)):
                yy, xx = y + dy, x + dx
                if 0 <= yy < h and 0 <= xx < w and not skip[yy, xx]:
                    nbrs.append(cfa[yy, xx])
            if len(nbrs) < 2:
                continue
            med = np.median(nbrs)
            if abs(cfa[y, x] - med) > threshold:
                cfa[y, x] = med
                count += 1
    return cfa, count


def fix_defect_region(cfa, region_cols, y_lo, y_hi, threshold=55.0):
    """楔形区域缺陷修复: 区域内像素与"带外同相位参考"差 > threshold 则替换。

    带外参考 = 左右最近非缺陷列 (同相位, x±2 步进) 的均值; 关键是从缺陷带外
    取参考, 避免缺陷像素互相污染。区域外像素原样保留。

    region_cols: 疑似缺陷列集合 (如 range(502, 517))。
    y_lo, y_hi: 检查行范围 (半开区间 [y_lo, y_hi))。
    返回 (修复后的副本, 替换掩码, 替换像素数)。
    """
    cfa = cfa.astype(np.float64).copy()
    h, w = cfa.shape
    replaced = np.zeros((h, w), dtype=bool)
    count = 0
    for y in range(y_lo, min(y_hi, h)):
        for x in range(w):
            if x not in region_cols:
                continue
            lv = rv = None
            lx = x - 2
            while lx >= 0:
                if lx not in region_cols:
                    lv = cfa[y, lx]
                    break
                lx -= 2
            rx = x + 2
            while rx < w:
                if rx not in region_cols:
                    rv = cfa[y, rx]
                    break
                rx += 2
            refs = [v for v in (lv, rv) if v is not None]
            if not refs:
                continue
            ref = np.mean(refs)
            if abs(cfa[y, x] - ref) > threshold:
                cfa[y, x] = ref
                replaced[y, x] = True
                count += 1
    return cfa, replaced, count


def render(cfa, pattern="RGGB", r_gain=None, b_gain=None, gamma=0.85, b_extra=0.93):
    """去马赛克 -> 灰度世界 WB -> 伽马, 返回 (h,w,3) uint8 RGB。

    cfa: 240x320 CFA (uint8 或 float64)。r_gain/b_gain 缺省时按全帧灰度世界
    自动计算 (R/B 增益使 R/G=B/G=1)。b_extra: 用户偏好的额外降蓝系数。
    """
    rgb = demosaic_bayer(cfa.astype(np.uint8), pattern).astype(np.float64)
    if r_gain is None:
        r_gain = rgb[:, :, 1].mean() / rgb[:, :, 0].mean()
    if b_gain is None:
        b_gain = rgb[:, :, 1].mean() / rgb[:, :, 2].mean()
    img = rgb.copy()
    img[:, :, 0] *= r_gain
    img[:, :, 2] *= b_gain * b_extra
    img = np.clip(img, 0, 255)
    img = 255 * (img / 255) ** gamma
    return img.astype(np.uint8)


def fix_leading_zeros(cfa):
    """修复每行前2个零字节 (Bayer pipeline delay): 用第2列数据填充前2列。

    OV7670 raw 模式每行前2字节始终为零 (传感器 pipeline delay)，
    不处理会导致图像横向拉伸 + 零值污染 demosaic/WB。
    返回修复后的副本 (原矩阵不变)。
    """
    cfa = cfa.copy()
    cfa[:, 0] = cfa[:, 2]
    cfa[:, 1] = cfa[:, 3]
    return cfa


def pipeline(raw, dead_threshold=60.0, fix_dead=True, pattern="RGGB",
             r_gain=None, b_gain=None, gamma=0.85, b_extra=0.93):
    """240x320 raw Bayer -> RGB: 修复零列 -> 散点死点修复 -> 去马赛克 -> WB -> 伽马。"""
    cfa = fix_leading_zeros(raw.astype(np.float64))
    stats = {}
    if fix_dead:
        cfa, stats["dead_replaced"] = fix_dead_pixels(cfa, dead_threshold)
    rgb = render(cfa, pattern, r_gain, b_gain, gamma, b_extra)
    return rgb, stats


def main(argv=None):
    p = argparse.ArgumentParser(
        description="OV7670 raw Bayer 320x240 管线: 死点修复 + 去马赛克 + 渲染 PNG")
    p.add_argument("input", help="240x320 raw 字节文件")
    p.add_argument("-o", "--output", required=True, help="输出 PNG")
    p.add_argument("--no-dead-fix", action="store_true",
                   help="跳过散点死点修复")
    p.add_argument("--dead-threshold", type=float, default=60.0,
                   help="死点判定阈值 (默认 %(default)s)")
    p.add_argument("--pattern", default="RGGB", help="CFA 模式 (默认 %(default)s)")
    p.add_argument("--r-gain", type=float, default=None,
                   help="R 增益 (默认全帧灰度世界)")
    p.add_argument("--b-gain", type=float, default=None,
                   help="B 增益 (默认全帧灰度世界)")
    p.add_argument("--b-extra", type=float, default=0.93,
                   help="额外降蓝系数 (默认 %(default)s)")
    p.add_argument("--gamma", type=float, default=0.85,
                   help="伽马 (默认 %(default)s)")
    args = p.parse_args(argv)

    raw = np.fromfile(args.input, dtype=np.uint8).reshape(240, 320)
    rgb, stats = pipeline(
        raw, dead_threshold=args.dead_threshold,
        fix_dead=not args.no_dead_fix,
        pattern=args.pattern,
        r_gain=args.r_gain, b_gain=args.b_gain,
        gamma=args.gamma, b_extra=args.b_extra)

    from PIL import Image
    Image.fromarray(rgb).save(args.output)
    print(f"{args.output}: {rgb.shape[1]}x{rgb.shape[0]} saved")
    if stats:
        print(f"  dead_replaced={stats.get('dead_replaced', 0)}")


if __name__ == "__main__":
    main()

File: bayer_pipeline.R
#!/usr/bin/env Rscript
# OV7670 Raw Bayer 320x240 pipeline: dead pixel fix -> demosaic -> WB -> gamma.
#
# Usage:
#   Rscript bayer_pipeline.R input.raw -o out.png
#   Rscript bayer_pipeline.R input.raw -o out.png --no-dead-fix

suppressPackageStartupMessages(library(png))

shift_matrix <- function(mat, dy, dx) {
    h <- nrow(mat)
    w <- ncol(mat)
    out <- matrix(0, h, w)
    src_rows <- (1:h) + dy
    src_cols <- (1:w) + dx
    rows_ok <- src_rows >= 1L & src_rows <= h
    cols_ok <- src_cols >= 1L & src_cols <= w
    if (any(rows_ok) && any(cols_ok)) {
        out[rows_ok, cols_ok] <- mat[src_rows[rows_ok], src_cols[cols_ok], drop = FALSE]
    }
    out
}

gather_avg <- function(src, h, w, target_mask, kernel) {
    vals <- matrix(0, h, w)
    cnt <- matrix(0, h, w)
    for (k in kernel) {
        dy <- k[1]
        dx <- k[2]
        shifted_src <- shift_matrix(src, dy, dx)
        shifted_tgt <- shift_matrix(target_mask, dy, dx)
        take <- shifted_tgt > 0
        vals <- vals + ifelse(take, shifted_src, 0)
        cnt <- cnt + take
    }
    vals / pmax(cnt, 1)
}

color_map <- function(pattern, h, w) {
    yy <- (0:(h - 1)) %% 2
    xx <- (0:(w - 1)) %% 2
    idx <- outer(yy * 2, xx, `+`) + 1L
    chars <- strsplit(pattern, "")[[1]]
    matrix(chars[idx], h, w)
}

demosaic_bayer <- function(cfa, pattern) {
    h <- nrow(cfa)
    w <- ncol(cfa)
    cfa_f <- cfa + 0.0
    cmap <- color_map(pattern, h, w)
    is_g <- matrix(as.numeric(cmap == "G"), h, w)
    is_r <- matrix(as.numeric(cmap == "R"), h, w)
    is_b <- matrix(as.numeric(cmap == "B"), h, w)

    kernel8 <- expand.grid(dy = -1:1, dx = -1:1)
    kernel8 <- kernel8[!(kernel8$dy == 0 & kernel8$dx == 0), , drop = FALSE]
    kernel8 <- lapply(seq_len(nrow(kernel8)), function(i) c(kernel8$dy[i], kernel8$dx[i]))

    g_avg <- gather_avg(cfa_f, h, w, is_g, kernel8)
    g <- ifelse(is_g == 1, cfa_f, g_avg)

    r_avg <- gather_avg(cfa_f, h, w, is_r, kernel8)
    g_at_r <- gather_avg(g, h, w, is_r, kernel8)
    r <- ifelse(is_r == 1, cfa_f, r_avg * (g / pmax(g_at_r, 1e-9)))

    b_avg <- gather_avg(cfa_f, h, w, is_b, kernel8)
    g_at_b <- gather_avg(g, h, w, is_b, kernel8)
    b <- ifelse(is_b == 1, cfa_f, b_avg * (g / pmax(g_at_b, 1e-9)))

    r <- pmin(255, pmax(0, r))
    g <- pmin(255, pmax(0, g))
    b <- pmin(255, pmax(0, b))
    arr <- array(0, dim = c(h, w, 3))
    arr[, , 1] <- floor(r)
    arr[, , 2] <- floor(g)
    arr[, , 3] <- floor(b)
    arr
}

fix_dead_pixels <- function(cfa, threshold = 60.0) {
    cfa <- cfa + 0.0
    h <- nrow(cfa)
    w <- ncol(cfa)
    count <- 0L
    if (h < 5 || w < 5) return(list(cfa = cfa, count = count))
    for (y in 3:(h - 2)) {
        for (x in 3:(w - 2)) {
            nbrs <- numeric(0)
            for (off in list(c(-2, 0), c(2, 0), c(0, -2), c(0, 2))) {
                yy <- y + off[1]
                xx <- x + off[2]
                if (yy >= 1L && yy <= h && xx >= 1L && xx <= w) {
                    nbrs <- c(nbrs, cfa[yy, xx])
                }
            }
            if (length(nbrs) < 2) next
            med <- median(nbrs)
            if (abs(cfa[y, x] - med) > threshold) {
                cfa[y, x] <- med
                count <- count + 1L
            }
        }
    }
    list(cfa = cfa, count = count)
}

render <- function(cfa, pattern = "RGGB", r_gain = NULL, b_gain = NULL,
                   gamma = 0.85, b_extra = 0.93) {
    rgb <- demosaic_bayer(cfa, pattern) + 0.0
    if (is.null(r_gain)) r_gain <- mean(rgb[, , 2]) / mean(rgb[, , 1])
    if (is.null(b_gain)) b_gain <- mean(rgb[, , 2]) / mean(rgb[, , 3])
    img <- rgb
    img[, , 1] <- img[, , 1] * r_gain
    img[, , 3] <- img[, , 3] * b_gain * b_extra
    img[, , 1] <- pmin(255, pmax(0, img[, , 1]))
    img[, , 2] <- pmin(255, pmax(0, img[, , 2]))
    img[, , 3] <- pmin(255, pmax(0, img[, , 3]))
    img <- 255 * (img / 255)^gamma
    floor(img)
}

fix_zero_columns <- function(cfa) {
    h <- nrow(cfa)
    w <- ncol(cfa)
    if (w < 2) return(cfa)
    for (r in seq_len(h)) {
        row <- cfa[r, ]
        i <- 1
        while (i <= w) {
            if (row[i] != 0) { i <- i + 1; next }
            lo <- i
            while (i <= w && row[i] == 0) i <- i + 1
            hi <- i - 1L
            left_val <- NA; left_idx <- lo - 1L
            while (left_idx >= 1L) {
                if (row[left_idx] != 0) { left_val <- row[left_idx]; break }
                left_idx <- left_idx - 1L
            }
            right_val <- NA; right_idx <- hi + 1L
            while (right_idx <= w) {
                if (row[right_idx] != 0) { right_val <- row[right_idx]; break }
                right_idx <- right_idx + 1L
            }
            if (!is.na(left_val) && !is.na(right_val)) {
                span <- right_idx - left_idx
                for (j in lo:hi) {
                    t_val <- (j - left_idx) / span
                    row[j] <- left_val * (1 - t_val) + right_val * t_val
                }
            } else if (!is.na(left_val)) {
                row[lo:hi] <- left_val
            } else if (!is.na(right_val)) {
                row[lo:hi] <- right_val
            }
        }
        cfa[r, ] <- row
    }
    cfa
}

pipeline <- function(raw, dead_threshold = 60.0, fix_dead = TRUE,
                     pattern = "RGGB",
                     r_gain = NULL, b_gain = NULL, gamma = 0.85, b_extra = 0.93) {
    cfa <- fix_zero_columns(raw + 0.0)
    stats <- list()
    if (fix_dead) {
        res <- fix_dead_pixels(cfa, dead_threshold)
        cfa <- res$cfa
        stats$dead_replaced <- as.integer(res$count)
    }
    rgb <- render(cfa, pattern, r_gain, b_gain, gamma, b_extra)
    list(rgb = rgb, stats = stats)
}

main <- function(argv) {
    args <- argv
    get_flag <- function(flag) {
        i <- match(flag, args)
        if (is.na(i)) FALSE else { args[i] <<- NA; TRUE }
    }
    get_val <- function(flag, default) {
        i <- match(flag, args)
        if (is.na(i)) default else { v <- args[i + 1L]; args[i] <<- NA; args[i + 1L] <<- NA; v }
    }

    inputs <- args[!is.na(args) & !grepl("^-", args)]
    if (length(inputs) < 1) stop("usage: Rscript bayer_pipeline.R input.raw -o out.png", call. = FALSE)
    input <- inputs[1]
    output <- get_val("-o", NULL)
    if (is.na(output)) output <- get_val("--output", NULL)
    if (is.na(output)) stop("-o/--output required", call. = FALSE)

    no_dead <- get_flag("--no-dead-fix")
    dead_threshold <- as.numeric(get_val("--dead-threshold", "60"))
    r_gain_str <- get_val("--r-gain", NA)
    b_gain_str <- get_val("--b-gain", NA)
    b_extra <- as.numeric(get_val("--b-extra", "0.93"))
    gamma <- as.numeric(get_val("--gamma", "0.85"))

    r_gain <- if (is.na(r_gain_str)) NULL else as.numeric(r_gain_str)
    b_gain <- if (is.na(b_gain_str)) NULL else as.numeric(b_gain_str)

    con <- file(input, "rb")
    on.exit(close(con))
    bytes <- readBin(con, "raw", n = 240 * 320)
    if (length(bytes) < 240 * 320) stop(sprintf("%s: %d B < need %d B (240x320)", input, length(bytes), 240 * 320), call. = FALSE)
    raw <- matrix(as.integer(bytes[1:(240 * 320)]), nrow = 240, ncol = 320, byrow = TRUE)

    res <- pipeline(raw, dead_threshold = dead_threshold, fix_dead = !no_dead,
                    r_gain = r_gain, b_gain = b_gain, gamma = gamma, b_extra = b_extra)

    writePNG(res$rgb / 255, target = output)
    cat(sprintf("%s: %dx%d saved\n", output, ncol(res$rgb), nrow(res$rgb)))
    st <- res$stats
    if (length(st) > 0) {
        cat(sprintf("  dead_replaced=%d\n", st$dead_replaced))
    }
    invisible(res)
}

if (sys.nframe() == 0 && !interactive()) {
    main(commandArgs(trailingOnly = TRUE))
}
